README.md
# WORK IN PROGRESS ! WORK IN PROGRESS !

> The `ppx_bitstring` project adds Erlang-style bitstrings and matching over
> bitstrings as a `ppx` syntax extension and library for OCaml.  It is directly
> based on a similar syntax extension for `camlp4` written by Richard
> W.M. Jones.

You can use this module to both parse and generate binary formats, files and
protocols.

Bitstring handling is added as primitives to the language, making it
exceptionally simple to use and very powerful.

Here is how to parse the header from a GIF image:

```ocaml
let bits = Bitstring.bitstring_of_file "image.gif" in
match%bitstring bits with
| ("GIF87a"|"GIF89a") [@l 6*8] [@string], (* GIF magic. *)
  width [@l 16] [@littleendian],
  height [@l 16] [@littleendian] ->
    printf "%s: GIF image is %d x %d pixels" filename width height
| _ ->
    eprintf "%s: Not a GIF image\n" filename
```

This example shows how to parse the first few fields in a Linux ELF binary (see
`<elf.h>`):

```ocaml
let bits = Bitstring.bitstring_of_file "/bin/ls" in
match%bitstring bits with
| 0x7f [@l 8], "ELF" [@l 24] [@string], (* ELF magic number *)
  e_ident [@l 12*8] [@bitstring],    (* ELF identifier *)
  e_type [@l 16] [@littleendian],    (* object file type *)
  e_machine [@l 16] [@littleendian]  (* architecture *)
  ->
    printf "This is an ELF binary, type %d, arch %d\n"
      e_type e_machine
```

And an IPv4 packet:

```ocaml
match%bitstring header with
| version [@l 4], hdrlen [@l 4], tos [@l 8], length [@l 16],
  identification [@l 16], flags [@l 3], fragoffset [@l 13],
  ttl [@l 8], protocol [@l 8], checksum [@l 16],
  source [@l 32],
  dest [@l 32],
  options [@l (hdrlen-5)*32] [@bitstring],
  payload [@l -1] [@bitstring]
  when version = 4 ->

    printf "IPv%d:\n" version;
    printf "  header length: %d * 32 bit words\n" hdrlen;
    printf "  type of service: %d\n" tos;
    printf "  packet length: %d bytes\n" length;
    printf "  identification: %d\n" identification;
    printf "  flags: %d\n" flags;
    printf "  fragment offset: %d\n" fragoffset;
    printf "  ttl: %d\n" ttl;
    printf "  protocol: %d\n" protocol;
    printf "  checksum: %d\n" checksum;
    printf "  source: %lx  dest: %lx\n" source dest;
    printf "  header options + padding:\n";
    Bitstring.hexdump_bitstring stdout options;
    printf "  packet payload:\n";
    Bitstring.hexdump_bitstring stdout payload
```

Bitstring handles integers, strings, sub-bitstrings, big-, little- and
native-endianness, signed and unsigned types, variable-width fields, fields with
arbitrary bit alignment.

# LICENSE

The library is licensed under the LGPL v2 or later, with the OCaml
linking exception.  See the file COPYING.LIB for full terms.

Programs are licensed under the GPL v2 or later.  See the file COPYING
for full terms.

All examples and tests are public domain.


CHANGES
------------------------------------------------------------------------
r197 | richard.wm.jones@gmail.com | 2012-08-10 12:52:44 +0100 (Fri, 10 Aug 2012) | 1 line
Changed paths:
   M /trunk/Makefile.in
   M /trunk/configure.ac

Stable OCaml dependencies.
------------------------------------------------------------------------
r196 | richard.wm.jones@gmail.com | 2012-08-10 12:44:18 +0100 (Fri, 10 Aug 2012) | 31 lines
Changed paths:
   M /trunk/bitstring_c.c

Fix memory leaks and alignment issues (thanks rixed at happyleptic.org)

First one: a huge memory leak when bitmatching on int64. The reason is
that on my arch (MIPS), which is a ARCH_ALIGN_INT64 arch, the function
to extract int64 calls caml_copy_int64 but you are not supposed to do
that since all functions are declared noalloc (should segfault but
merely leaks memory on my program). See this message of Xavier for
more explanations as to why it's a bug:

http://caml.inria.fr/pub/ml-archives/caml-list/2010/03/2e386c384c04f4a424acc44a1
fae3abd.en.html

So the easy fix would be to remove the noallocs for the int64 functions,
but there is a better way around the issue: instead of assigning a int64
onto a possibly non 8 bytes aligned memory location we can merely memcpy
the value in there, thus avoiding the extra allocation and the
caml_c_call wrapper.

While looking at this I stumbled upon the other bug: bitstring_c.c
extract values by assigning a casted rvalue into a properly
stack-allocated local variable. You cannot do that, this is undefined
behavior since it assumes that unaligned reads are permitted. Again, on
my MIPS such unaligned access are (by default) diverted to an exception
handler which (slowly!) fix them on the fly (one can see the errorcount in
/proc/kernel/debug/mips/unaligned_instructions), but on other archs it
could as well result in a sigbus. I believe this would be a problem on
ARM as well. Again, memcpy comes to the rescue.

The attached patch fixes both issues.


------------------------------------------------------------------------
r195 | richard.wm.jones@gmail.com | 2012-07-20 09:06:08 +0100 (Fri, 20 Jul 2012) | 3 lines
Changed paths:
   M /trunk/META.in

META: Add -parser and -printer options.


------------------------------------------------------------------------
r194 | richard.wm.jones@gmail.com | 2012-04-04 10:40:42 +0100 (Wed, 04 Apr 2012) | 4 lines
Changed paths:
   M /trunk/bitstring.ml

takebits/dropbits: check that n >= 0
http://code.google.com/p/bitstring/issues/detail?id=18


------------------------------------------------------------------------
r193 | richard.wm.jones@gmail.com | 2012-03-30 13:49:30 +0100 (Fri, 30 Mar 2012) | 4 lines
Changed paths:
   M /trunk/bitstring.ml

Permissive subbitstring allows a segmentation fault (issue #16).
Fix by mrvn.


------------------------------------------------------------------------
r192 | richard.wm.jones@gmail.com | 2012-01-17 13:24:46 +0000 (Tue, 17 Jan 2012) | 3 lines
Changed paths:
   M /trunk/MANIFEST

Update MANIFEST.


------------------------------------------------------------------------
r191 | richard.wm.jones@gmail.com | 2012-01-17 13:21:57 +0000 (Tue, 17 Jan 2012) | 3 lines
Changed paths:
   M /trunk/.depend
   M /trunk/configure.ac

Version 2.0.3.


------------------------------------------------------------------------
r190 | richard.wm.jones@gmail.com | 2012-01-17 13:20:53 +0000 (Tue, 17 Jan 2012) | 5 lines
Changed paths:
   M /trunk/bitstring.ml
   A /trunk/tests/test_91_concat.ml

Fix concatenation of non-aligned bitstrings (thanks Phil Tomson).

This commit includes a regression test.


------------------------------------------------------------------------
r189 | richard.wm.jones@gmail.com | 2012-01-17 13:02:18 +0000 (Tue, 17 Jan 2012) | 5 lines
Changed paths:
   M /trunk/pa_bitstring.ml
   A /trunk/tests/test_90_bind_as.ml

In a '{...} as foo' binding, save the original bitstring offset and
length as 'foo'.  Before we were saving the running offset and length.
This commit also adds a regression test.  (Bug found by Matej Kosik).


------------------------------------------------------------------------
r188 | richard.wm.jones@gmail.com | 2012-01-17 12:44:36 +0000 (Tue, 17 Jan 2012) | 3 lines
Changed paths:
   M /trunk/Makefile.in
   M /trunk/README

Add 'make check' as a synonym for 'make test'.


------------------------------------------------------------------------
r187 | richard.wm.jones@gmail.com | 2012-01-17 12:39:09 +0000 (Tue, 17 Jan 2012) | 3 lines
Changed paths:
   D /trunk/tests/01_load.ml
   D /trunk/tests/02_run.ml
   D /trunk/tests/10_match_bits.ml
   D /trunk/tests/11_match_ints.ml
   D /trunk/tests/15_extract_int.ml
   D /trunk/tests/18_extract_32_64_int.ml
   D /trunk/tests/20_varsize.ml
   D /trunk/tests/30_bitbuffer.ml
   D /trunk/tests/31_bitstring_concat.ml
   D /trunk/tests/32_bitstring_compare.ml
   D /trunk/tests/33_substring.ml
   D /trunk/tests/34_takebits.ml
   D /trunk/tests/35_load_from_file.ml
   D /trunk/tests/40_endianexpr.ml
   D /trunk/tests/50_named_pattern.ml
   D /trunk/tests/51_open_pattern.ml
   D /trunk/tests/60_simple_offset.ml
   D /trunk/tests/61_offset_string.ml
   D /trunk/tests/62_offset_padding.ml
   D /trunk/tests/65_save_offset_to.ml
   D /trunk/tests/70_check_and_bind.ml
   D /trunk/tests/80_hexdump.ml
   A /trunk/tests/test_01_load.ml (from /trunk/tests/01_load.ml:186)
   A /trunk/tests/test_02_run.ml (from /trunk/tests/02_run.ml:186)
   A /trunk/tests/test_10_match_bits.ml (from /trunk/tests/10_match_bits.ml:186)
   A /trunk/tests/test_11_match_ints.ml (from /trunk/tests/11_match_ints.ml:186)
   A /trunk/tests/test_15_extract_int.ml (from /trunk/tests/15_extract_int.ml:186)
   A /trunk/tests/test_18_extract_32_64_int.ml (from /trunk/tests/18_extract_32_64_int.ml:186)
   A /trunk/tests/test_20_varsize.ml (from /trunk/tests/20_varsize.ml:186)
   A /trunk/tests/test_30_bitbuffer.ml (from /trunk/tests/30_bitbuffer.ml:186)

bitstring_persistent.ml
(* Bitstring persistent patterns.
 * Copyright (C) 2008 Red Hat Inc., Richard W.M. Jones
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version,
 * with the OCaml linking exception described in COPYING.LIB.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 *
 * $Id: bitstring_persistent.ml 151 2008-08-26 08:21:43Z richard.wm.jones $
 *)

open Printf

open Longident
open Asttypes
open Parsetree
open Ast_helper

type patt = Parsetree.pattern
type expr = Parsetree.expression
type loc_t = Location.t

(* Field.  In bitmatch (patterns) the type is [patt field].  In
 * BITSTRING (constructor) the type is [expr field].
 *)
type 'a field = {
  field : 'a;                           (* field ('a is either patt or expr) *)
  flen : expr;                          (* length in bits, may be non-const *)
  endian : endian_expr;                 (* endianness *)
  signed : bool;                        (* true if signed, false if unsigned *)
  t : field_type;                       (* type *)
  loc : Location.t;                     (* location in source code *)
  offset : expr option;                 (* offset expression *)
  check : expr option;                  (* check expression [patterns only] *)
  bind : expr option;                   (* bind expression [patterns only] *)
  save_offset_to : patt option;         (* save_offset_to [patterns only] *)
}
and field_type = Int | String | Bitstring (* field type *)
and endian_expr =
  | ConstantEndian of Bitstring.endian  (* a constant little/big/nativeendian *)
  | EndianExpr of expr                  (* an endian expression *)

type pattern = patt field list

type constructor = expr field list

type named = string * alt
and alt =
  | Pattern of pattern
  | Constructor of constructor

(* Work out if an expression is an integer constant.
 *
 * Returns [Some i] if so (where i is the integer value), else [None].
 *
 * Fairly simplistic algorithm: we can only detect simple constant
 * expressions such as [k], [k+c], [k-c] etc.
 *)
let rec expr_is_constant = function
  | {pexp_desc = Pexp_constant (Const_int i)} ->      (* Literal integer constant. *)
    Some i
  | [%expr [%e? a] + [%e? b]] -> (* Addition of constants. *)
    (match expr_is_constant a, expr_is_constant b with
     | Some a, Some b -> Some (a+b)
     | _ -> None)
  | [%expr [%e? a] - [%e? b]] -> (* Subtraction. *)
    (match expr_is_constant a, expr_is_constant b with
     | Some a, Some b -> Some (a-b)
     | _ -> None)
  | [%expr [%e? a] * [%e? b]] -> (* Multiplication. *)
    (match expr_is_constant a, expr_is_constant b with
     | Some a, Some b -> Some (a*b)
     | _ -> None)
  | [%expr [%e? a] / [%e? b]] -> (* Division. *)
    (match expr_is_constant a, expr_is_constant b with
     | Some a, Some b -> Some (a/b)
     | _ -> None)
  | [%expr [%e? a] lsl [%e? b]] -> (* Shift left. *)
    (match expr_is_constant a, expr_is_constant b with
     | Some a, Some b -> Some (a lsl b)
     | _ -> None)
  | [%expr [%e? a] lsr [%e? b]] -> (* Shift right. *)
    (match expr_is_constant a, expr_is_constant b with
     | Some a, Some b -> Some (a lsr b)
     | _ -> None)
  | _ -> None                           (* Anything else is not constant. *)

let string_of_field_type = function
  | Int -> "int"
  | String -> "string"
  | Bitstring -> "bitstring"

let patt_printer = function
  | { ppat_desc = Ppat_var { txt = id } } -> id
  | { ppat_desc = Ppat_any } -> "_"
  | _ -> "[pattern]"

let rec expr_printer = function
  | { pexp_desc = Pexp_ident { txt = Lident id } } -> id
  | { pexp_desc = Pexp_constant (Const_int i) } -> string_of_int i
  | { pexp_desc = Pexp_apply ({ pexp_desc = Pexp_ident { txt = Lident op } }, [(_, a); (_, b)]) } ->
    sprintf "%s %s %s" op (expr_printer a) (expr_printer b)
  | _ -> "[expr]"

let _string_of_field { flen = flen;
                       endian = endian; signed = signed; t = t;
                       loc = loc;
                       offset = offset; check = check; bind = bind;
                       save_offset_to = save_offset_to } =
  let flen = expr_printer flen in
  let endian =
    match endian with
    | ConstantEndian endian -> Bitstring.string_of_endian endian
    | EndianExpr expr -> sprintf "endian(%s)" (expr_printer expr) in
  let signed = if signed then "signed" else "unsigned" in
  let t = string_of_field_type t in

  let offset =
    match offset with
    | None -> ""
    | Some expr -> sprintf ", offset(%s)" (expr_printer expr) in

  let check =
    match check with
    | None -> ""
    | Some expr -> sprintf ", check(%s)" (expr_printer expr) in

  let bind =
    match bind with
    | None -> ""
    | Some expr -> sprintf ", bind(%s)" (expr_printer expr) in

  let save_offset_to =
    match save_offset_to with
    | None -> ""
    | Some patt ->
        match patt with
        | { ppat_desc = Ppat_var { txt = id } } -> sprintf ", save_offset_to(%s)" id
        | _ -> sprintf ", save_offset_to([patt])" in

  let loc_fname = loc.Location.loc_start.Lexing.pos_fname in
  let loc_line = loc.Location.loc_start.Lexing.pos_lnum in
  let loc_char = loc.Location.loc_start.Lexing.pos_cnum - loc.Location.loc_start.Lexing.pos_bol in

  sprintf "%s : %s, %s, %s%s%s%s%s (* %S:%d %d *)"
    flen t endian signed offset check bind save_offset_to
    loc_fname loc_line loc_char

let rec string_of_pattern_field ({ field = patt } as field) =
  sprintf "%s : %s" (patt_printer patt) (_string_of_field field)

and string_of_constructor_field ({ field = expr } as field) =
  sprintf "%s : %s" (expr_printer expr) (_string_of_field field)

let string_of_pattern pattern =
  "{ " ^
    String.concat ";\n  " (List.map string_of_pattern_field pattern) ^
    " }\n"

let string_of_constructor constructor =
  "{ " ^
    String.concat ";\n  " (List.map string_of_constructor_field constructor) ^
    " }\n"

let named_to_channel chan n = Marshal.to_channel chan n []

let named_to_string n = Marshal.to_string n []

let named_to_buffer str ofs len n = Marshal.to_buffer str ofs len n []

let named_from_channel = Marshal.from_channel

let named_from_string = Marshal.from_string

let create_pattern_field loc =
  {
    field = Pat.any ();
    flen = Exp.constant (Const_int 32);
    endian = ConstantEndian Bitstring.BigEndian;
    signed = false;
    t = Int;
    loc = loc;
    offset = None;
    check = None;
    bind = None;
    save_offset_to = None;
  }

let set_lident_patt field id =
  let loc = field.loc in
  { field with field = Pat.var ~loc { txt = id; loc } }
let set_int_patt field i =
  let loc = field.loc in
  { field with field = Pat.constant ~loc (Const_int i) }
let set_string_patt field str =
  let loc = field.loc in
  { field with field = Pat.constant ~loc (Const_string (str, None)) }
let set_unbound_patt field =
  let loc = field.loc in
  { field with field = Pat.any ~loc () }
let set_patt field patt = { field with field = patt }
let set_length_int field flen =
  let loc = field.loc in
  { field with flen = Exp.constant ~loc (Const_int flen) }
let set_length field flen = { field with flen = flen }
let set_endian field endian = { field with endian = ConstantEndian endian }
let set_endian_expr field expr = { field with endian = EndianExpr expr }
let set_signed field signed = { field with signed = signed }
let set_type_int field = { field with t = Int }
let set_type_string field = { field with t = String }
let set_type_bitstring field = { field with t = Bitstring }
let set_location field loc = { field with loc = loc }
let set_offset_int field i =
  let loc = field.loc in
  { field with offset = Some (Exp.constant ~loc (Const_int i)) }
let set_offset field expr = { field with offset = Some expr }
let set_no_offset field = { field with offset = None }
let set_check field expr = { field with check = Some expr }
let set_no_check field = { field with check = None }
let set_bind field expr = { field with bind = Some expr }
let set_no_bind field = { field with bind = None }
let set_save_offset_to field patt = { field with save_offset_to = Some patt }
let set_save_offset_to_lident field id =
  let loc = field.loc in
  { field with save_offset_to = Some (Pat.var ~loc { txt = id; loc }) }
let set_no_save_offset_to field = { field with save_offset_to = None }

let create_constructor_field loc =
  {
    field = Exp.constant (Const_int 0);
    flen = Exp.constant (Const_int 32);
    endian = ConstantEndian Bitstring.BigEndian;
    signed = false;
    t = Int;
    loc = loc;
    offset = None;
    check = None;
    bind = None;
    save_offset_to = None;
  }

let set_lident_expr field id =
  let loc = field.loc in
  { field with field = Exp.ident ~loc { txt = Lident id; loc } }
let set_int_expr field i =
  let loc = field.loc in
  { field with field = Exp.constant ~loc (Const_int i) }
let set_string_expr field str =
  let loc = field.loc in
  { field with field = Exp.constant ~loc (Const_string (str, None)) }
let set_expr field expr =
  let loc = field.loc in
  { field with field = expr }

let get_patt field = field.field
let get_expr field = field.field
let get_length field = field.flen
let get_endian field = field.endian
let get_signed field = field.signed
let get_type field = field.t
let get_location field = field.loc
let get_offset field = field.offset
let get_check field = field.check
let get_bind field = field.bind
let get_save_offset_to field = field.save_offset_to
